Provide a wrapper for simple submodule clone steps

tiennou opened the pull request:
While working on the submodule-in-worktree issue, I had to finagle with #4149. So let's add that to the API.

Arguably, it could be named `git_submodule_add` and perform all the steps by itself.

pks-t commented on src/submodule.c:
```diff
@@ -832,6 +832,47 @@ int git_submodule_repo_init(
+	opts.remote_cb = clone_return_origin;
+	opts.remote_cb_payload = submodule;
+
+	error = git_clone(&clone, git_submodule_url(submodule), git_submodule_path(submodule), &opts);
```
I'm surprised this actually works in the test suite. The problem is that when I've just initialized a submodule via `git_submodule_add_setup`, it will already contain the ".git" gitlink to the actual gitdir of the submodule. Because of that, `git_clone` will refuse to clone into an non-empty directory. And in fact it does fail when changing my example code to use this function instead of implementing the clone step myself. Puzzled

pks-t replied:
Never mind, now it works. I think the examples just didn't build. And in fact the `clone_return_repo` callback should be the reason why it succeeds

pks-t replied:
Argh. I've managed to hard-reset my adjustment to use your new API, so never mind my never mind comment. It does fail in `git_clone`, line 406, where it checks whether the local path is an empty directory or not.

pks-t replied:
Ah, I got it. `git_submodule_path` returns a relative path, which is why `git_clone` thinks that the target path doesn't yet exist. You have to create an absolute path for submodule relative to the parent repo's working directory, which will subsequently break your test as `git_clone` will error out due to files existing in the target directory.

Sorry for my confused comments

tiennou commented on src/submodule.c:
```diff
@@ -832,6 +832,47 @@ int git_submodule_repo_init(
+	return git_submodule_open(out, sm);
+}
+
+int git_submodule_add_clone(git_repository **out, git_submodule *submodule)
```
I'd be tempted to say that the "old" way already provides this, and shrug it off. I'm adding this API for the 95% of users that just want to `git submodule add` something and don't care for wacky remote configurations.

pks-t replied:
This function only clones and doesn't do the add, right? Wouldn't `git_submodule_clone` be easier to understand in that case?

tiennou replied:
I… hadn't anticipated it would make sense 😆. My thinking was that, being the step between `add_setup` and `add_finalize`, it made sense to have it under that namespace. If it so happens that it can be used to clone actual uncloned submodules, then why not, but that wasn't the plan (and I'm eyeing towards the unimplemented checkout actions 😉).

pks-t replied:
So, uh. Can it be used to clone uncloned submodules right now? If so, I'd rename it. Otherwise it's fine as is, I guess.

tiennou commented:
Now I don't know what changes you're requesting anymore 😉.

pks-t commented:
Sorry for my comments being confusing :P

The thing is that your function unfortunately doesn't work. The test you've written works by accident due to a bug in `git_submodule_add_clone`, as you clone into `git_submodule_path(submodule)`, which returns a path relative to the parent repo. So the `git_clone` will in fact clone into the current working directory and not into the parent repo. What you have to do is to make `git_submodule_path(submodule)` an absolute path relative to the parent repo and then do the clone. This is where the next issue will surface: as the submodule directory already exists and due to a ".git" gitlink file existing inside of it, the `git_clone` step will error out due to the target directory not being empty.

tiennou commented:
Okay, I fixed the function, so yeah the test wasn't working as expected because of pathing issues (which I wasn't expecting so, good catch 😉).

There's two ways to fix : either the clone existence check is wrong, and this should be changed, or it's not, and the clone must be performed manually.

Now for the discussion part, is clone wrong ? We're using the "approved" `clone_opts->repository_cb` function for grabbing a user-provided repository, but the emptiness check happens first. I can think of a few workarounds : 
- trust that the user gave us something sane via the `repository_cb`, in that it would have called `git_repository_open` to be able to do that. The definition of sane in that case is fuzzy though.
- make the empty check more precise by saying that empty allows a valid gitlink at the root of the folder, but that looks to be a deviation from git (I've checked against a manually crafted gitlink). So I'd make it a 👎.

pks-t commented:
I don't like the second option, it's much too specific. While I think the first option is quite nice, I don't know if we really want to relax the check like that. So, option number three:

- create a permit-clone-into-a-populated-directory flag (defaulting to `false`) in `git_clone_options`, and if it is set  we skip the check for emptiness

tiennou commented:
I've gone a 4th way: add a private variant of `git_clone` with a relaxed check and use that. That restricts it to this specific function call, which is sufficient, and keep it from being exposed in the API (less API to hang you with).

pks-t commented on src/clone.c:
```diff
@@ -382,11 +382,12 @@ int git_clone__should_clone_local(const char *url_or_path, git_clone_local_t loc
 }
 
-int git_clone(
+static int git__clone(
```
`clone` would be sufficient, I guess :)

tiennou replied:
I think you got me there, I just had the Linux buildbots remind me that it's some syscall 🤣.

pks-t commented on src/submodule.c:
```diff
@@ -815,6 +816,54 @@ int git_submodule_repo_init(
+	opts.remote_cb_payload = submodule;
+
+	git_buf_puts(&rel_path, git_repository_workdir(git_submodule_owner(submodule)));
+	git_buf_joinpath(&rel_path, git_buf_cstr(&rel_path), git_submodule_path(submodule));
```
These two need error checking.

pks-t commented on tests/clone/nonetwork.c:
```diff
@@ -353,44 +353,19 @@ void test_clone_nonetwork__clone_from_empty_sets_upstream(void)
-	clone_opts.remote_cb = just_return_origin;
-	clone_opts.remote_cb_payload = sm;
-	cl_git_pass(git_clone(&sm_repo, cl_fixture("testrepo.git"), "testrepo", &clone_opts));
+	cl_git_pass(git_submodule_add_clone(NULL, sm));
```
I think we should have a new test that uses `git_submodule_add_clone`, as we should retain the test that verifies the old way of doing that.

tiennou commented:
Apart from the naming question, this is good to go (my reply somehow doesn't show up from the conversation view though).

tiennou commented:
Rebased, I've renamed the function to `git_submodule_clone`, and I've moved its tests to submodule/add.c (it seemed more relevant). I hope that's fine.

pks-t commented on include/git2/submodule.h:
```diff
@@ -285,6 +286,20 @@ GIT_EXTERN(int) git_submodule_add_setup(
+ */
+GIT_EXTERN(int) git_submodule_clone(
+	git_repository **out,
+	git_submodule *submodule);
```
To not repeat past mistakes, do we maybe want to introduce an options struct here from the get-go? Doesn't need to contain anything options for now

tiennou replied:
I usually like having a default "do the default thing" and provide an `_ext` on top of that, if/when the need arises. But it feels a little premature (though only because of a lack of use cases, so if you have a definite not too difficult "this is needed" example, I'll reconsider).

ethomson replied:
Hmm.  I would very quickly want progress reporting if I were to use this in anger.  I concede that this is a bit tricky.  I would definitely want, eg, progress reporting but many fields should not be eligible to be set by an end-user, so taking a whole `git_clone_options` may not be appropriate.  Additionally, we may need to do some proxy'ing with some of the callbacks to ensure that payloads are delivered correctly.

Having said that, progress reporting and cancellation still feels like table-stakes for any function that does network I/O.

pks-t replied:
> I usually like having a default "do the default thing" and provide an _ext on top of that, if/when the need arises.

Well, defaults are easily attainable by just passing `NULL`. Personally, I'm not too happy with the `_ext` suffix and I see it rather as a "Okay we f'd up the first time, so let's just tack on a suffix and fix up the function".

pks-t replied:
And yeah, fully agree with the usecases @ethomson proposes.

tiennou replied:
I've looked at it a little, and effectively, just accepting a `git_clone_options` seems a bit too much, and since you could just do the clone yourself, a more limited but easier to use API seems nicer. I'm just not sure how to balance what to expose, especially if it requires another struct 😟.

So, from the clone struct, we're not using :
- `git_checkout_options`: loads of things, but apart from the notify/progress/prefdata, I don't think anything else makes sense.
- `git_remote_options`: `callbacks` looks like a given, not sure about the rest, but I suspect it would be better if this one was accessible.
- `bare`, `local` : I don't think those need to be exposed (though maybe `local` ?)
- `checkout_branch`: sounds related, but I think that's not up to the user at this point ? ie, the submodule should have been setup correctly in the first place.

```c
typedef struct {
/*…*/
 git_checkout_(notify|progress|pref)_cb $step_checkout_cb;
 void *checkout_payload;
 git_remote_options remote_callbacks;
} git_submodule_clone_options;
```

All those callbacks make me want to audit them and see if they couldn't be make more generic 😭.

ethomson replied:
Hmm, yeah that's a tough decision.  I can see pros and cons to simply taking a `git_clone_options` struct and just ignoring and overwriting parts of it (documenting this of course).  Or a simpler struct and proxying that data into a `git_clone_options`.

The former seems like the simpler strategy for you (the implementer).  What I'm not sure of is whether it's easier for users of the library, which is of course where I think the important bits are.

pks-t replied:
Simply taking `git_clone_options` is not going to fly in case there's going to be submodule-cloning specific options at some point in time, which I'd say is not all that unlikely. So I think it should definitely be a new `git_submodule_clone_options` struct, but as you said it may just embed and forward (parts of) the `git_clone_options` struct.

tiennou replied:
I've noticed that `git_submodule_update_options` seemed to have mostly what we needed, so I've added this as our option struct. Sorry, I forgot to rebase-ping, but IIRC it was ready to 🚢 as well.

pks-t replied:
Oh, didn't see that, sorry.

`git_submodule_update_options` does look similar to what we need, that's true. I'm still a bit hesitant, though. It includes options that cannot fit the bill right now and we may need to extend the struct for either `git_submodule_update` or `git_submodule_clone` with options that are incompatible with the respective other function. If you ask me, that's trading ease of implementation (no need for an additional struct) for a confusing API. I'm more in the camp of using one struct per function whenever it's not obvious that they're always going to take the same set of opts.

Just my 2 cents -- I'm happy to be overruled here due to my recent lack of time for libgit2.

tiennou replied:
(Struct for reference)
```c
typedef struct git_submodule_update_options {
	unsigned int version;
	git_checkout_options checkout_opts;
	git_fetch_options fetch_opts;
	int allow_fetch;
} git_submodule_update_options;
```

Of theses, `allow_fetch` feels the most dubious, since we don't have any repository before the call. `checkout` seems [fine](https://libgit2.org/libgit2/#HEAD/type/git_checkout_options) (though I think there's a few rough edges), and `fetch` looks [okay](https://libgit2.org/libgit2/#HEAD/type/git_fetch_options).

Now, I don't have that much experience with submodules, but a shared "those are the settings to obey if an update needs to happen", especially given the `git submodule update --init --recursive` thingie seems to be a nice thing. Writing in the docblock that `allow_fetch` only makes sense if the submodule is already initialized, and ignoring it if it's set seems sufficient.

pks-t commented:
Thanks, @tiennou! Looks good to me except for the one comment I have

pks-t commented:
Rebased your branch to fix conflicts.

pks-t commented:
I've removed the homegrown clone test, which didn't do what one expected (the resulting submodule was cloned outside of the parent repo) and added two new tests

tiennou commented:
Thanks for keeping up with this ! ✨
